### PR TITLE
Export peering cli

### PR DESCRIPTION
**Pull request by nathancoleman**
### Description
Describe why you're making this change, in plain English.

### Testing & Reproduction steps
* In the case of bugs, describe how to replicate
* If any manual tests were done, document the steps and the conditions to replicate
* Call out any important/ relevant unit tests, e2e tests or integration tests you have added or are adding

### Links
Include any links here that might be helpful for people reviewing your PR (Tickets, GH issues, API docs, external benchmarks, tools docs, etc). If there are none, feel free to delete this section.

Please be mindful not to leak any customer or confidential information. HashiCorp employees may want to use our internal URL shortener to obfuscate links.

### PR Checklist

* [ ] updated test coverage
* [ ] external facing docs updated
* [ ] not a security concern

### CODE REVIEW

File: command/peering/export/export.go
Comment (nathancoleman) [inline command/peering/export/export.go:38]: @DanStough we were thinking it might make sense to use the `namespace` flag inherited via `flags.Merge(c.flags, c.http.ClientFlags())` (and maybe the partition flag inherited the same way) instead of defining a new flag here. Thoughts?

File: command/peering/export/export.go
Comment (nathancoleman) [inline command/peering/export/export.go:135]: Notably, we can assign a namespace here even when using OSS. Discussed in the issue w/ @DanStough that we will want to update the config-entry validation on the server to return an error such as `enterprise-only feature`.

File: command/peering/export/export.go
Comment (nathancoleman) [inline command/peering/export/export.go:185]: Might be worth including multiple peers in the example, `-peers=cluster-1,cluster-2`

**Review by nathancoleman — COMMENTED** [review]
One more question for you @DanStough: should we be making API calls to query for the service being exported to ensure that it exists, or should we depend on the server to validate that, or is that up to the user?

File: command/peering/export/export.go
Comment (DanStough) [reply command/peering/export/export.go:38]: It doesn't look like we use that flag for anything else relevant to storing the `exported-services`, but it will automatically get added as a query parameter in the request: https://github.com/hashicorp/consul/blob/c046d1a4d870639227baff629ff304a1b72deede/api/api.go#L1019-L1021

No idea how Consul will handle that. You could test it out or use a dedicated field to be safe.

**Comment by DanStough** [comment]
@nathancoleman , w.r.t. checking if the service exists, I would say no. Generally, design things in the CLI to support eventual consistency, so you don't need to worry about the order in which these things are created.

**Comment by DanStough** [comment]
This looks awesome! One note is this would look almost identical if we made a command for exporting to partitions. I wonder if this should be a new top-level command, and we can create a followup tasks to support both `-peers` and `-partitions`?

**Comment by nathancoleman** [comment]
@DanStough the question of whether this single command supports both peers and partitions or whether we have this command and another parition-specific command elsewhere in the CLI is something I'm expecting an answer from Jared on before we merge

**Comment by nathancoleman** [comment]
@DanStough if you could please review this, it'd be super appreciated!

We've split things out such that OSS here only has the `-name` and `-consumer-peers` flags.

The `-namespace`, `-partition` and `-consumer-partitions` flags will be added in followup work based on our discussion yesterday.

File: command/services/export/export.go
Comment (DanStough) [inline command/services/export/export.go:191]: nit: I think there some naunce between `cluster name`, which confused with datacenter and `peer name`, which is a local idea.
```suggestion
Usage: consul services export [options] -name <service name> -consumer-peers <peer names>

  Export a service. The peer names provided will be used locally by
  this cluster to refer to the peer where the services will be exported. 
```

File: command/services/export/export_oss.go
Comment (DanStough) [inline command/services/export/export_oss.go:20]:
```suggestion
	c.flags.StringVar(&c.peerNames, "consumer-peers", "", "(Required) A list of peers to export the service to, formatted as a comma-separated list.")
```

File: command/services/export/export.go
Comment (DanStough) [inline command/services/export/export.go:97]: Maybe this comparison should move to its own function in `export_oss.go`? That way for enterprise, we can also check the namespace.

File: command/services/export/export.go
Comment (DanStough) [inline command/services/export/export.go:85]: Optional: instead of the many levels of nesting here, I might move the logic in the if statement out into helper functions like `CreateExportConfigEntry` and `UpdateExportConfigEntry`.

File: command/services/export/export_test.go
Comment (DanStough) [inline command/services/export/export_test.go:58]: I assume this is only true for OSS, so it would need to go in a file `export_test_oss.go`.

**Review by DanStough — COMMENTED** [review]
Overall looks great! A few small things and one big question: Do we need to have the OSS/ENT functionality separated?  I'm looking at [service register](https://github.com/hashicorp/consul/blob/main/command/services/register/register.go#L64), and you can see we include all of the flags there. Usually, we rely on the server endpoints to correctly reject enterprise-y things if it is open source.

File: command/services/export/export_oss.go
Comment (DanStough) [inline command/services/export/export_oss.go:43]: I don't think this split makes sense anymore that all the flags and much of the logic are now in `export.go`. I would consolidate everything into that file.

**Review by DanStough — CHANGES_REQUESTED** [review]
I think we can finish getting rid of the `export_oss.go` go logic since all the flags and some of the logic moved into the shared file.

Did you want to include docs in this PR? There is an empty `export.mdx` file. Your call if you want to add that here or in a follow-up.

File: command/services/export/export_test.go
Comment (DanStough) [inline command/services/export/export_test.go:100]: In a perfect world, we would move the partition test cases to another file `export_test_ent.go`, but since it looks like we don't have any server-side validation for this, I'm fine leaving it as it is.

**Review by DanStough — APPROVED** [review]
🥳 🍾 🎆 :shipit:

File: command/services/export/export.go
Comment (jkirschner-hashicorp) [inline command/services/export/export.go:62]: What happens if partitions are specified in OSS?

File: command/services/export/export.go
Comment (jkirschner-hashicorp) [inline command/services/export/export.go:106]: Nice due diligence in handling this case :)

File: command/services/export/export.go
Comment (jkirschner-hashicorp) [inline command/services/export/export.go:110]: Should we write "cluster peers" instead of "peers"?

I ask because, in the context of HCP, there's also "HVN peering/peers".

Same comment applies to the case immediately below.

(I could go either way)

File: command/services/export/export.go
Comment (jkirschner-hashicorp) [inline command/services/export/export.go:150]: Is there any other validation of peer or partition names beyond "is it empty" that the backend does?

I guess the possible harm is that someone could write a config entry for a non-existent partition, or one with a mistyped/garbled name?

That said, we output a string like `Successfully exported service %q to peers %q` afterwards, so that would hopefully show whether something was wrong.

So maybe additional validation isn't important (unless it was really easy to do).

**Review by jkirschner-hashicorp — COMMENTED** [review]
Left a few optional suggestions for wording clarifications and some questions. Looks good - nice work team!

File: command/services/export/export.go
Comment (nathancoleman) [reply command/services/export/export.go:150]: @DanStough and I talked at one point about adding validation in the backend for this as a separate PR.

Today, you can write a config entry that exports to a non-existent peer or partition using `consul config write` - this command is no different since it uses the same config entry mechanism. Based on that, I consider adding more validation on the server side to be outside the scope of this PR. It would be good to get on someone's radar though. Thoughts?

File: command/services/export/export.go
Comment (nathancoleman) [reply command/services/export/export.go:110]: I like the additional clarity. Will update

File: .changelog/15654.txt
Comment (nathancoleman) [reply .changelog/15654.txt]: Reworded to talk about "adding a command for" instead of "adding support for"

File: command/services/export/export.go
Comment (nathancoleman) [reply command/services/export/export.go:62]: This follows the same theme as other enterprise features in this PR. The command with partitions specified will go through even for OSS users, same as if you use `$ consul config write` today (since this is really all that's happening behind the scenes).

Since the established pattern is to allow Enterprise features in the OSS CLI, the correct thing to do in the longer term IMO is to add server validation for this config entry that returns an error in this case.

**Review by jkirschner-hashicorp — APPROVED** [review]
LGTM :)

**Comment by david-yu** [comment]
@nathancoleman Out of curiosity are there docs changes for this features as well? I currently do not see docs for this feature here: https://developer.hashicorp.com/consul/commands/services

**Comment by nathancoleman** [comment]
@david-yu good catch - meant to get docs done before the 1.16 release and completely forgot. I'll get a PR up today and backport in 1.16.x.

**Pull request merged**
